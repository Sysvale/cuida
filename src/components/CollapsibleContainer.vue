<template>
	<span id="collapsible-container">
		<div
			class="collapsible-container__item"
			@click="triggerExpanded"
		>
			<span
				:class="internalValue ? 'collapsible-container__chevron--opened' : 'collapsible-container__chevron--closed'"
			/>
			<!-- @slot Slot para renderização do conteúdo do header/title customizado do item
			-->
			<slot name="title">
				<span
					class="collapsible-container__title"
				>
					{{ title }}
				</span>
			</slot>
		</div>
		<div
			v-if="internalValue"
			class="collapsible-container__content"
			data-testid="collapsible-content"
		>
			<!-- @slot Slot para renderização conteúdo do container
				quando o componente está expandido
			-->
			<slot />
		</div>
	</span>
</template>

<script>
export default {
	name: 'CdsCollapsibleContainer',
	props: {
		/**
		 * Representa o estado do componente. Quando 'false' o conteúdo não é mostrado
		 * e quando 'true' o conteúdo do container é exibido.
		 */
		modelValue: {
			type: Boolean,
			default: false,
			required: true,
		},
		/**
		 * Título do item a ser expandido
		 */
		title: {
			type: String,
			default: 'Visualizar mais',
		},
	},

	data() {
		return {
			internalValue: this.modelValue,
		};
	},

	watch: {
		internalValue(newValue) {
			/**
			* Evento emitido quando o conteúdo do item é exibido ('true') ou ocultado ('false').
			* @event update:model-value
			* @type {Event}
			*/
			this.$emit('update:modelValue', newValue);
		},

		modelValue(newValue) {
			this.internalValue = newValue;
		},
	},

	methods: {
		triggerExpanded() {
			this.internalValue = !this.internalValue;
		},
	},
};
</script>

<style lang="scss">
@use '../assets/sass/tokens/index' as tokens;

.collapsible-container {
	&__content {
		padding: tokens.px(5);

		& p {
			margin: tokens.my(1);
		}
	}

	&__chevron--closed {
		top: 6px;
		right: 2px;
		display: block;
		height: 32px;
		width: 32px;
		pointer-events: none;
		border-radius: tokens.$border-radius-small;
		display: flex;
		align-items: center;
		justify-content: center;

		&::before,
		&::after {
			content: '';
			width: 0;
			height: 0;
		}

		&::before {
			border-left: 6px solid transparent;
			border-right: 6px solid transparent;
			border-bottom: 6px solid tokens.$n-400;
			border-radius: tokens.$border-radius-small;
			transition: all 300ms ease-in-out;
			transform: rotate(-180deg);
		}
	}

	&__chevron--opened {
		top: 6px;
		right: 2px;
		display: block;
		height: 32px;
		width: 32px;
		pointer-events: none;
		border-radius: tokens.$border-radius-small;
		display: flex;
		align-items: center;
		justify-content: center;

		&::before,
		&::after {
			content: '';
			width: 0;
			height: 0;
		}

		&::before {
			border-left: 6px solid transparent;
			border-right: 6px solid transparent;
			border-bottom: 6px solid tokens.$n-400;
			border-radius: tokens.$border-radius-small;
			transition: all 300ms ease-in-out;
			transform: rotate(0deg);
		}
	}

	&__item {
		font-size: 14px;
		color: tokens.$n-600;
		display: flex;
		align-items: start;
		cursor: pointer;
		display: flex;
		align-items: baseline;
		margin: tokens.ml(n2);

		& hr {
			border-color: tokens.$n-600 !important;
		}
	}

	&__title {
		font-weight: 700;
	}

	&__icon--collapsed {
		color: tokens.$n-600;
		margin: tokens.mr(2);
		transition: all 0.25s ease-in-out;
	}

	&__icon--expanded {
		@extend .collapsible-container__icon--collapsed;
		transform: rotate(180deg);
	}
}
</style>
