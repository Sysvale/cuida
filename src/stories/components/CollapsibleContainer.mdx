import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import CollapsibleContainer from '../../components/CollapsibleContainer.vue';
import * as CollapsibleContainerStories from './CollapsibleContainer.stories';

<Meta of={CollapsibleContainerStories} />

export const Template = (args) => ({
	components: { CdsCollapsibleContainer: CollapsibleContainer },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-collapsible-container
			v-bind="args" 
		>
			<p class="p-3">
				Mussum Ipsum, cacilds vidis litro abertis. Todo mundo
				vê os porris que eu tomo, mas ninguém vê os tombis que eu levo!
				Atirei o pau no gatis, per gatis num morreus. Pra lá , depois
				divoltis porris, paradis. Leite de capivaris, leite de mula
				manquis sem cabeça. Viva Forevis aptent taciti sociosqu ad
				litora torquent. Per aumento de cachacis, eu reclamis. Nec
				orci ornare consequat. Praesent lacinia ultrices consectetur.
				Sed non ipsum felis. Cevadis im ampola pa arma uma pindureta.
			</p>
		</cds-collapsible-container>
	`,
});

# CollapsibleContainer

### CollapsibleContainers são componentes utilizados para reduzir o espaço vertical quando há uma grande quantidade de informações, com a possibilidade de ser expandido para exibir o conteúdo sumarizado.

***

<br />

## Quando usar:

* For necessário otimizar o espaço vertical.
* Se deseja mostrar itens sumarizados.
* Se possui uma grande quantidade de informações que podem ser agrupadas em um tópico/título.

<br />

## Quando não usar:

* Houver poucos itens a ser sumarizados;

<br />

## Preview

<Canvas>
  <Story of={CollapsibleContainerStories.CollapsibleContainer} />
</Canvas>

<Controls />
