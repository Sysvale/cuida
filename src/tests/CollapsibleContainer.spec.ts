import { describe, test, expect } from 'vitest';
import CollapsibleContainer from '../components/CollapsibleContainer.vue';
import { flushPromises, mount } from '@vue/test-utils';
  
describe('CollapsibleContainer', () => {
	test('renders correctly', async () => {
		const wrapper = mount(CollapsibleContainer, {
			props: {
				modelValue: true,
				title: 'Seu universo expandido',
			},
		});
  
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('content is rendered correctly on modelValue change', async () => {
		expect.assertions(2);

		const wrapper = mount(CollapsibleContainer, {
			props: {
				modelValue: false,
				title: 'Seu universo expandido',
			},
		});
  
		expect(wrapper.findAll('[data-testid="collapsible-content"]').length).toBe(0);

		wrapper.setProps({
			modelValue: true,
		});

		await flushPromises();

		expect(wrapper.findAll('[data-testid="collapsible-content"]').length).toBe(1);
	})
});
