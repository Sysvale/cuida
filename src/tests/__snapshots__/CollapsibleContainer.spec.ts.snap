// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollapsibleContainer > renders correctly 1`] = `
"<span id="collapsible-container"><div class="collapsible-container__item"><span class="collapsible-container__chevron--opened"></span>
<!-- @slot Slot para renderização do conteúdo do header/title customizado do item
			--><span class="collapsible-container__title">Seu universo expandido</span></div>
<div class="collapsible-container__content" data-testid="collapsible-content">
  <!-- @slot Slot para renderização conteúdo do container
				quando o componente está expandido
			-->
</div></span>"
`;
